### ISSUE TITLE
Enderal normal map issue

### ISSUE THREAD

**Issue by Hortophyll** [hakasapl/ParallaxGen#20]
As discussed, here is the request regarding my problem with the normal maps in enderal.

enderal uses its own meshes with its own diffuse textures, but many of them have normal maps from skyrim in use. 

this leads to the problem for my complex parallax mod, which is currently in progress, that with such meshes not my texture is patched, but that of a skyrim mod. 

as can be seen in the following example

![iIriOvl](https://github.com/user-attachments/assets/8e660f22-3f17-4d67-b5fb-40dfc5378884)

the solution would be to either manually adjust all normal maps and make the meshes available. or ideally i can use your tool for this. i specify the path to the directory of my textures and your tool matches the diffuse textures with the normal maps and patches all affected meshes. my normal maps always have an _n.dds at the end.

i could then make these meshes available and users of enderal could then use your program as normal to patch the remaining meshes as they wish

if you need more information, please contact me, discord is also fine

thank you

**Comment by hakasapl** [comment]
I believe this would be best as a `pgtools` function. If you didn't know `pgtools` is a CLI tool designed for mod developers that will have a bunch of useful function related to parallaxgen. I believe this function could fit there. pgtools is slated to be available sometime in the 0.6 release cycle, so sometime soon.

**Comment by Hortophyll** [comment]
Wow nice! A big thank you :)

Do you think there is a problem in the following case: diffuse:  example_d.dds  or  example_diff.dds

My normal map: example_d_n.dds  or  example_diff_n.dds

And similarly strange names of diffuse textures?

**Comment by Hortophyll** [comment]
Ok, so if I understand this correctly, I have the new problem that in the end my normal maps no longer match the diffuse ones if they are named as in my examples. Am I right? 


Or, I go through the textures and look at all the suffixes, make a list and give it to you to see what else you could add?   Then all suffixes would be removed and it would fit together?

**Comment by hakasapl** [comment]
Yeah, though now I'm second guessing how I implemented this so maybe I can figure out a better solution that doesn't break in these cases

**Comment by Hortophyll** [comment]
I wish I could contribute more :/

But then let's do it this way, you tell me when I should do what exactly and then I do it :D

**Comment by hakasapl** [comment]
In the current 0.6 build of parallaxgen, you have to use standard suffixes only. So _d or _n (but not _d_n or _diff etc.).

I will try to figure out a better way to do this for a future version but I don't have guidance on that yet because I'm not sure what it will look like

**Comment by Hortophyll** [comment]
Yes, that's exactly why I meant, I don't have any stress myself because I still have to tweak textures. Whatever the solution would be then, that's how I'll do it :) 

I also sent you a small donation, generally as a thank you for your programme and for wanting to help me. That is not a matter of course :)

**Comment by hakasapl** [comment]
@Hortophyll thank you, I appreciate it! I will keep you updated

**Comment by Hortophyll** [comment]
I thought about it a bit and wondered if the solution would be to patch enderal meshes with an old version of parallaxgen so that the cm maps are set. 

Then it would need a parallaxgen function that does the same with the normal maps (check the diffuse and simple rename the normal maps accordingly : diffuse: enderal.dds Normal: Skyrim_n.dds becomes enderal_n.dds) and the whole thing ignores suffixes except "_n"

So I would have patched all enderal meshes.

I have already created texture sets myself in a plugin, so parallaxgen could then only take care of all skyrim specific things 

Problem is, I don't know if it's easy or that simple to do, that's up to you to decide :)

**Comment by Hortophyll** [comment]
Hello again :)

A friend of mine wants to help me and tries to write something so that I have patched all normal maps in correct form in the nifs.

I will let you know if there are any news about this :)

**Comment by hakasapl** [comment]
Sorry I missed your earlier comment. Just to be clear - are you planning on shipping meshes with your mod with these changes or need parallaxgen to do these things for the user?

**Comment by Hortophyll** [comment]
No problem :) 

If hopefully everything works, I would provide the meshes. I don't see any possibility for the user to do that. That would be too complicated 



but to be clear, only with complex material enabled for my enderal textures. This means that there are still many meshes that contain additional skyrim textures. 

And for the most part, enderal also uses standard skyrim meshes

To patch these it is necessary to use parallaxgen afterwards. 

If my meshes have already partially activated cm with matching normal map, parallaxgen would only patch the rest, right? Or do I still have to worry about the suffixes?



(assuming the user has additional cm/parallax mods installed)

**Comment by hakasapl** [comment]
Yeah, parallaxgen only patches what it needs to so if it's already correct it won't re-save it

**Comment by Hortophyll** [comment]
Super! Then that's a good plan, I think. I hope you're ok with it?

And I hope my friend in helping hand gets it done :)

**Comment by Hortophyll** [comment]
@hakasapl Hello good sir, at the weekend the friend I mentioned wrote me a small mesh patcher. This allowed me to customise all the normal maps in the meshes as described. In the end there were about 550 meshes. 

A first test with version 6 delivered good results as expected and everything should work now

I still have 2 questions about upcoming versions

On the one hand I noticed in a mesh that where the double sided flag was set it was patched anyway. Is this still planned? 

And secondly, as far as I know, the environmental map scale will be set to 1 in the mesh. Does this always happen or can I make exceptions?

Thank you

**Comment by hakasapl** [comment]
> @hakasapl Hello good sir, at the weekend the friend I mentioned wrote me a small mesh patcher. This allowed me to customise all the normal maps in the meshes as described. In the end there were about 550 meshes.
> 
> A first test with version 6 delivered good results as expected and everything should work now

Cool!

> I still have 2 questions about upcoming versions
> 
> On the one hand I noticed in a mesh that where the double sided flag was set it was patched anyway. Is this still planned?

I haven't looked into double sided stuff that much - does it break to patch it?

> And secondly, as far as I know, the environmental map scale will be set to 1 in the mesh. Does this always happen or can I make exceptions?

It will be set to 1 only for CM patches, nothing else. It's assumed that if you are making CM textures that you are making your env mask conform to env map scale of 1 in the NIF (that's generally the convention CM authors have followed and not setting it to 1 causes a bunch of issues). But let me know if you think that's not right

**Comment by Hortophyll** [comment]
The thing with double sided is that one side of it always distorts when moving and therefore, as far as I know, they should not be patched (for cm at least) so for example banners or there is a market stand in enderal where the textile roof is double sided


I have already thought about the env map scale, but I have a few metal textures that are used for a mesh that is in good condition and one that is already weathered. 

So I have set a different value for them to achieve the effect of weathering

But if all the shaders for cm and the corresponding textures are already set for this part of the mesh, does your tool change anything at all?

**Comment by Hortophyll** [comment]
Ok i see it does patch them anyway :)

Now I have to learn how to exclude meshes with a JSON. But there are only a few, so no problem :D

Is it possible to specify that only 1 Bstrishape in a mesh is not patched?

I come back to my market stand, which has Bstrishape that should be patched (vanilla skyrim textures) and the one fabric that has double sided and should not be patched.

But this texture is also used elsewhere where it is not double sided in the mesh, so I would have liked to keep it

One question I have about the latest version, I have installed a portable version of mo2 with enderal se as modversion, based on skyrim AE. 

I can't specify the path in the gui. (the instance ini is directly in the portable MO2 folder) This is not a problem for me as my modlist is sorted, I just wanted to point it out

All in all, I'm very excited because I'm very close to publication :D

**Comment by Hortophyll** [comment]
Please A little instruction on how to enter the json correctly would be very helpful. I'm not familiar with it and I've just spent ages with it only that at the end of the patch process all my entries are gone again because something was wrong somewhere :(

**Comment by Hortophyll** [comment]
> One question I have about the latest version, I have installed a portable version of mo2 with enderal se as modversion, based on skyrim AE.
> 
> I can't specify the path in the gui. (the instance ini is directly in the portable MO2 folder) This is not a problem for me as my modlist is sorted, I just wanted to point it out
> 

read it on nexus in the comments, it works now :D sorry!


> Please A little instruction on how to enter the json correctly would be very helpful. I'm not familiar with it and I've just spent ages with it only that at the end of the patch process all my entries are gone again because something was wrong somewhere :(

i got it to work, but why is it deleted again after one use?

**Comment by hakasapl** [comment]
Sorry, I need to figure out why my GH notifications aren't working. What gets deleted again? You are editing user.json right?

**Comment by hakasapl** [comment]
Ah, you're right. The user.json does get deleted. I will fix that in the 0.7.1 hotfix coming soon.

You'll be able to set blocklist etc. in the GUI in a future update, probably 0.7.2 if I had to guess.

As for only patching specific parts of a NIF, that would be hard to implement properly right now. There will be a day where mod authors can include custom jsons with their mod to change PG behavior like that, but that feature has not begun development at all. After discussing with some other authors most of them said not patching double sided at all would break a lot of their textures, so the only way is to properly implement that json system I think

**Comment by Hortophyll** [comment]
All right, master, I know my way around!

I'll stick to what the others say, I'm just a little modder. Maybe that's why I sometimes come up with strange solutions that in hindsight weren't so clever after all xD 

Thank you for your very, very long patience with me :D

What else I noticed, with me now with 2 times use of PG, esps were loaded into the parallaxgen esp with which are not activated at all in mo2

Maybe I still have to adjust something

In any case, it would be good if you could select enderalse as the game in the modersion and the enderal  bsa archives would then be loaded correctly 

If you like I will send you which ones they are, but wildscribe had told you about it on the parallax nexus page

Otherwise, many thanks for everything

**Comment by hakasapl** [comment]
Yeah can you send me those plugins? Are you saying the enderal plugins don't get loaded correctly?

**Comment by hakasapl** [comment]
Going to close this - correct me if I'm wrong but you have a solution for this? If not I can reopen
